File: N12761.java
import java.io.*;
import java.util.*;

public class N12761 {

    static int A, B, N, M;
    static int[] map;

    public static void main(String[] args) throws IOException {
        read();
        solution();
        write();
    }

    public static void read() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        A = Integer.parseInt(st.nextToken());
        B = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        br.close();
    }

    public static void solution() {
        map = new int[100001];
        Arrays.fill(map, Integer.MAX_VALUE);
        Queue<Integer> q = new LinkedList<Integer>();
        map[N] = 0;
        q.add(N);

        while (!q.isEmpty()) {
            int cur = q.poll();
            if (map[cur] >= map[M]) continue;
            int[] next = {cur + 1, cur - 1, cur + A, cur + B, cur - A, cur - B, cur * A, cur * B};
            for (int pos : next) {
                if (inRange(pos) && map[cur] + 1 < map[pos]) {
                    map[pos] = map[cur] + 1;
                    q.add(pos);
                }
            }
        }
    }

    public static void write() throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(String.valueOf(map[M]));
        bw.flush();
        bw.close();
    }

    public static boolean inRange(int x) {
        return 0 <= x && x <= 100000;
    }
}

File: N12761.kt
import java.io.*;
import java.util.*;

class Node(val t: Int, val p: Int) : Comparable<Node> {
    var time: Int = t
    var pos: Int = p

    override fun compareTo(other: Node): Int {
        return this.time - other.time
    }
}

private val bw = BufferedWriter(OutputStreamWriter(System.`out`))
var arr = IntArray(100001, { i -> Int.MAX_VALUE })
var a = 0
var b = 0
var n = 0
var m = 0

fun main(args: Array<String>) {
    init()
    solution()
}

private fun init() {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val st = StringTokenizer(br.readLine())
    a = st.nextToken().toInt()
    b = st.nextToken().toInt()
    n = st.nextToken().toInt()
    m = st.nextToken().toInt()
}

private fun solution() {
    var pq = PriorityQueue<Node>()

    pq.add(Node(0, n))
    arr[n] = 0

    while (!pq.isEmpty()) {
        var cur: Node = pq.poll()
        if (cur.pos == m) {
            bw.write(cur.time.toString())
            break
        }
        for (x in nextPos(cur.pos)) {
            if (inRange(x) && cur.time + 1 < arr[x]) {
                arr[x] = cur.time + 1
                pq.add(Node(cur.time + 1, x))
            }
        }
    }

    bw.flush()
    bw.close()
}

fun nextPos(pos: Int): IntArray {
    return intArrayOf(pos + 1, pos - 1, pos * a, pos * b, pos + a, pos + b, pos - a, pos - b)
}

fun inRange(pos: Int): Boolean {
    return pos >= 0 && pos <= 100000
}
